File: medium/2958_length_of_longest_subarray_with_at_most_k_frequency/solution.cpp
#include <unordered_map>
#include <vector>

class Solution {
public:
  int maxSubarrayLength(std::vector<int> &nums, int k) {
    std::unordered_map<int, int> numsCounter;

    size_t left = 0;
    size_t right = 0;

    size_t maxLen = 0;

    for (size_t left{}, right{}; right < nums.size();) {
      int num = nums[right];

      if (numsCounter[num] < k) {
        ++numsCounter[num];
        ++right;
      } else {
        while (numsCounter[num] == k)
          --numsCounter[nums[left++]];
      }

      maxLen = std::max(maxLen, right - left);
    }

    return static_cast<int>(maxLen);
  }
};

File: medium/2958_length_of_longest_subarray_with_at_most_k_frequency/solution.py
from collections import defaultdict


class Solution:
    def maxSubarrayLength(self, nums: list[int], k: int) -> int:
        n = len(nums)

        numsCounter = defaultdict(int)

        left = 0
        right = 0

        maxLen = 0

        while right < n:
            num = nums[right]

            if numsCounter[num] < k:
                numsCounter[num] += 1
                right += 1
            else:
                while numsCounter[num] == k:
                    numsCounter[nums[left]] -= 1
                    left += 1

            maxLen = max(maxLen, right - left)

        return maxLen
